### PR TITLE
Avoid `pandas` constructors in `dask.dataframe.core`

### PR DESCRIPTION
**Pull request by rjzamora**
There are many places in dask.dataframe where `pd.DataFrame`/`pd.Series` constructors are used explicitly. This PR proposes the addition of `serial_frame_constructor` and `serial_series_constrictor` utilities that take in an optional `like` parameter to determine which `DataFrame`/`Series` constructor to use (i.e. `pandas` or `cudf`). Default is `pandas.DataFrame` and `pandas.Series`.

The optional `like` argument is currently expected to be a serial `DataFrame`, serial `Series`, `DataFrame` collection, or `Series` collection.  It may also make sense to handle a numpy/cupy Array or Array collection (along the lines of @quasiben's suggestion in [#11889](https://github.com/rapidsai/cudf/issues/11889)).  Howeer, that feature will probablty require the addition of a new `array_to_frame` diispatch as well (or something similar).

- [x] Closes [#11889](https://github.com/rapidsai/cudf/issues/11889)
- [ ] Tests added / passed
- [ ] Passes `pre-commit run --all-files`

### CODE REVIEW

**Comment by quasiben** [comment]
That's a very cool solution to this problem -- and mirror's what @pentschev did for [NEP37](https://numpy.org/neps/nep-0037-array-module.html)

File: dask/dataframe/utils.py [hunk 1]
```diff
@@ -734,3 +734,23 @@ def drop_by_shallow_copy(df, columns, errors="raise"):
...
     """NotImplementedError and AttributeError"""
+
+
+def serial_frame_constructor(like=None):
+    """Return a serial DataFrame constructor"""
+    if is_dask_collection(like):
+        like = like._meta
+    if hasattr(like, "to_frame"):
+        # `like` is a Series rather than a DataFrame
+        like = like.iloc[:1].to_frame()
```
Comment (jrbourbeau) [inline dask/dataframe/utils.py:745]: Would `is_series_like(...)` and `Series._constructor_expanddim` work here instead?

File: dask/dataframe/core.py [hunk 2]
```diff
@@ -7150,7 +7159,7 @@ def cov_corr_combine(data_in, corr=False):
     return out
 
 
-def cov_corr_agg(data, cols, min_periods=2, corr=False, scalar=False):
+def cov_corr_agg(data, cols, min_periods=2, corr=False, scalar=False, like_df=None):
```
Comment (jrbourbeau) [inline dask/dataframe/core.py:7162]: What is `data` here? Can we get the DataFrame type from that?

**Review by jrbourbeau — COMMENTED** [review]
Thanks @rjzamora. Just left a few quick comments while passing by, I'll take a more detailed look soon

File: dask/dataframe/utils.py [hunk 3]
```diff
@@ -734,3 +734,23 @@ def drop_by_shallow_copy(df, columns, errors="raise"):
 
 class AttributeNotImplementedError(NotImplementedError, AttributeError):
     """NotImplementedError and AttributeError"""
+
+
+def serial_frame_constructor(like=None):
```
[the remark this answers is not in the document]
Comment (rjzamora) [reply dask/dataframe/utils.py:739]: Sure, makes sense to me

File: dask/dataframe/utils.py [hunk 4]
```diff
@@ -734,3 +734,23 @@ def drop_by_shallow_copy(df, columns, errors="raise"):
...
+    """Return a serial DataFrame constructor"""
+    if is_dask_collection(like):
```
[the remark this answers is not in the document]
Comment (rjzamora) [reply dask/dataframe/utils.py:747]: `_Frame._constructor` seems to return `new_dd_object` (I *think*). So, a bit different from what we want.

File: dask/dataframe/core.py (hunk 2 above)
Comment (rjzamora) [reply dask/dataframe/core.py:7162]: `data` is `list[dict[str, np/cupy.ndarray]]` here. Therfore, we would need to extend `serial_frame_constructor` to handle array-like data if we want to avoid the `like_df` argument.

File: dask/dataframe/utils.py [hunk 5]
```diff
@@ -734,3 +736,53 @@ def drop_by_shallow_copy(df, columns, errors="raise"):
...
+        except AttributeError:
+            like = like.meta
+    if is_dataframe_like(like):
+        return like._constructor_sliced
+    elif is_series_like(like):
+        return like._constructor
+    elif is_index_like(like):
+        return like.to_frame()._constructor_sliced
+    elif is_arraylike(like):
+        return serial_constructor_from_array(like, series=True)
```
Comment (jrbourbeau) [inline dask/dataframe/utils.py:786]: Including conversion from arrays adds an extra layer of complexity. Why is this needed?

File: dask/dataframe/tests/test_dataframe.py [hunk 6]
```diff
@@ -3202,6 +3202,24 @@ def test_cov():
     assert res._name != res3._name
 
 
+@pytest.mark.gpu
+def test_cov_gpu():
```
[the remark this answers is not in the document]
Comment (rjzamora) [reply dask/dataframe/tests/test_dataframe.py:3206]: Perhaps. We would need to change the `_compat` dataframe-creation machinery to use dispachable functions to do this.

File: dask/dataframe/utils.py (hunk 5 above)
Comment (rjzamora) [reply dask/dataframe/utils.py:786]: >Why is this needed?

It is not **needed** for this PR. I added it as an illustration for what would be needed to avoid passing the `like_df` argument ([see your comment here](https://github.com/dask/dask/pull/9570#discussion_r994767598)).  With that said, I do think it could be usefult for dask.dataframe to have a mechanism for creating a "serial" dataframe from array-like data.

File: dask/dataframe/utils.py [hunk 7]
```diff
@@ -734,3 +736,53 @@ def drop_by_shallow_copy(df, columns, errors="raise"):
...
+        return like._constructor_expanddim
+    elif is_index_like(like):
+        return like.to_frame()._constructor
+    elif is_arraylike(like):
+        return serial_constructor_from_array(like)
+    else:
+        raise TypeError(f"{type(like)} not supported by serial_frame_constructor")
+
+
+def serial_series_constructor(like):
```
[the remark this answers is not in the document]
Comment (rjzamora) [reply dask/dataframe/utils.py:766]: >Thoughts on the name of these methods?

I don't love the `serial_*` names but naming is indeed hard. I'm open to anything.

File: dask/dataframe/utils.py (hunk 5 above)
Comment (rjzamora) [reply dask/dataframe/utils.py]: **Update**: I removed array handling and the related dispatch from this PR

File: dask/dataframe/utils.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (rjzamora) [reply dask/dataframe/utils.py]: **Update**: Changed the names to `meta_*_constructor`

**Comment by rjzamora** [comment]
>I imagine it would be fairly easy for me or others to accidentally introduce a pandas constructor when one of these utilities would be more appropriate.

Yea, I also struggled to come up with a clever solutions for this problem (beyond the systematic expansion and maintenance of gpuci). One challenge is that `pd.DataFrame`/`Series` usage is not *always* a problem. In fact, the cudf codebase uses pandas in many places.  The primary issue is when a function inadvertently switches from cudf to pandas by assuming the input is pandas-based.  This specific case seems a bit difficult to detect in an automated way.

**Review by jrbourbeau — COMMENTED** [review]
Circling back here, @rjzamora are there any additional changes you'd like to include here? It'd be nice to include this in the release tomorrow if feasible

File: dask/dataframe/core.py [hunk 8]
```diff
@@ -7775,7 +7782,7 @@ def to_datetime(arg, meta=None, **kwargs):
 def to_timedelta(arg, unit=None, errors="raise"):
     if not PANDAS_GT_110 and unit is None:
         unit = "ns"
-    meta = pd.Series([pd.Timedelta(1, unit=unit)])
+    meta = meta_series_constructor(arg)([pd.Timedelta(1, unit=unit)])
```
Comment (jrbourbeau) [inline dask/dataframe/core.py:7785]: (Not a blocking comment, just meant for other reviewers) Noting that this will fail for some inputs (e.g. `numpy.ndarray`s) that are supported by `pandas.to_timedelta`. However, it looks like things already fail with `dd.to_timedelta` for such inputs on `main` and we arguably get a better error message with the changes in this PR. 

On `main`:

```python
In [1]: import numpy as np

In [2]: import dask.dataframe as dd

In [3]: dd.to_timedelta(np.arange(5), unit='s')
---------------------------------------------------------------------------
IndexError                                Traceback (most recent call last)
Input In [3], in <cell line: 1>()
----> 1 dd.to_timedelta(np.arange(5), unit='s')

File ~/projects/dask/dask/dask/dataframe/core.py:7779, in to_timedelta(arg, unit, errors)
   7777     unit = "ns"
   7778 meta = pd.Series([pd.Timedelta(1, unit=unit)])
-> 7779 return map_partitions(pd.to_timedelta, arg, unit=unit, errors=errors, meta=meta)

File ~/projects/dask/dask/dask/dataframe/core.py:6668, in map_partitions(func, meta, enforce_metadata, transform_divisions, align_dataframes, *args, **kwargs)
   6665     if collections:
   6666         simple = False
-> 6668 divisions = _get_divisions_map_partitions(
   6669     align_dataframes, transform_divisions, dfs, func, args, kwargs
   6670 )
   6672 if has_keyword(func, "partition_info"):
   6673     partition_info = {
   6674         (i,): {"number": i, "division": division}
   6675         for i, division in enumerate(divisions[:-1])
   6676     }

File ~/projects/dask/dask/dask/dataframe/core.py:6711, in _get_divisions_map_partitions(align_dataframes, transform_divisions, dfs, func, args, kwargs)
   6707 """
   6708 Helper to get divisions for map_partitions and map_overlap output.
   6709 """
   6710 if align_dataframes:
-> 6711     divisions = dfs[0].divisions
   6712 else:
   6713     # Unaligned, dfs is a mix of 1 partition and 1+ partition dataframes,
   6714     # use longest divisions found
   6715     divisions = max((d.divisions for d in dfs), key=len)

IndexError: list index out of range
``` 

With this PR:

```python
In [4]: dd.to_timedelta(np.arange(5), unit='s')
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
Input In [4], in <cell line: 1>()
----> 1 dd.to_timedelta(np.arange(5), unit='s')

File ~/projects/dask/dask/dask/dataframe/core.py:7785, in to_timedelta(arg, unit, errors)
   7783 if not PANDAS_GT_110 and unit is None:
   7784     unit = "ns"
-> 7785 meta = meta_series_constructor(arg)([pd.Timedelta(1, unit=unit)])
   7786 return map_partitions(pd.to_timedelta, arg, unit=unit, errors=errors, meta=meta)

File ~/projects/dask/dask/dask/dataframe/utils.py:782, in meta_series_constructor(like)
    780     return like.to_frame()._constructor_sliced
    781 else:
--> 782     raise TypeError(f"{type(like)} not supported by meta_series_constructor")

TypeError: <class 'numpy.ndarray'> not supported by meta_series_constructor
```

File: dask/dataframe/utils.py [hunk 9]
```diff
@@ -734,3 +734,49 @@ def drop_by_shallow_copy(df, columns, errors="raise"):
...
     """NotImplementedError and AttributeError"""
+
+
+def meta_frame_constructor(like):
+    """Return a serial DataFrame constructor
+
+    Parameters
+    ----------
+    like :
+        Any series-like or dataframe-like object.
```
Comment (jrbourbeau) [inline dask/dataframe/utils.py:745]: Index-like too

File: dask/dataframe/utils.py [hunk 10]
```diff
@@ -734,3 +734,49 @@ def drop_by_shallow_copy(df, columns, errors="raise"):
...
+        raise TypeError(f"{type(like)} not supported by meta_frame_constructor")
+
+
+def meta_series_constructor(like):
+    """Return a serial Series constructor
+
+    Parameters
+    ----------
+    like :
+        Any series-like or dataframe-like object.
```
Comment (jrbourbeau) [inline dask/dataframe/utils.py:768]: Index-like too

File: dask/dataframe/utils.py [hunk 11]
```diff
@@ -734,3 +734,49 @@ def drop_by_shallow_copy(df, columns, errors="raise"):
 
 class AttributeNotImplementedError(NotImplementedError, AttributeError):
     """NotImplementedError and AttributeError"""
+
+
+def meta_frame_constructor(like):
```
Comment (jrbourbeau) [inline dask/dataframe/utils.py:739]: Can we add some basic test coverage for `meta_frame_constructor` / `meta_series_constructor` to `dask/dataframe/tests/test_utils_dataframe.py`? In particular it'd be good to test that, for `meta_frame_constructor`,  `pd.DataFrame` / `pd.Series` / `pd.Index` return `pd.DataFrame` and we raise a `TypeError` when we expect (e.g. when a `np.ndarray` in used as input). Same for their gpu counterparts.

File: dask/dataframe/core.py [hunk 12]
```diff
@@ -2813,7 +2815,7 @@ def quantile(self, q=0.5, axis=0, numeric_only=True, method="default"):
                 graph = HighLevelGraph.from_collections(
                     keyname, layer, dependencies=quantiles
                 )
-                return DataFrame(graph, keyname, meta, quantiles[0].divisions)
+                return new_dd_object(graph, keyname, meta, quantiles[0].divisions)
```
Comment (jrbourbeau) [inline dask/dataframe/core.py:2818]: Nice -- thanks for cleaning up

File: dask/dataframe/utils.py (hunk 11 above)
Comment (rjzamora) [reply dask/dataframe/utils.py:739]: Sounds good!  Just added.

File: dask/dataframe/tests/test_utils_dataframe.py [hunk 13]
```diff
@@ -618,3 +620,33 @@ def check_custom_scheduler(part: pd.DataFrame) -> pd.DataFrame:
...
+        pd.Series([0]),
+        pd.Index([0]),
+        dd.from_dict({"x": [0]}, npartitions=1),
+        dd.from_dict({"x": [0]}, npartitions=1).x,
+        dd.from_dict({"x": [0]}, npartitions=1).index,
+    ],
+)
+def test_meta_constructor_utilities(data):
+    assert meta_series_constructor(data) == pd.Series
+    assert meta_frame_constructor(data) == pd.DataFrame
```
Comment (jrbourbeau) [inline dask/dataframe/tests/test_utils_dataframe.py:638]: Minor nit, but overall this looks good. Thanks again for adding this test

```suggestion
    assert meta_series_constructor(data) is pd.Series
    assert meta_frame_constructor(data) is pd.DataFrame
```
